### ISSUE TITLE
Move to NixOS org?

### ISSUE THREAD

**Issue by Ericson2314** [svanderburg/node2nix#206]
If https://github.com/NixOS/npm2nix is long deprecated, maybe that repo should be archived and this moved to NixOS to make it clear this the replacement? I only just learned 3 years after the fact that node2nix one :).

**Comment by svanderburg** [comment]
@Ericson2314 interesting discussion point. In addition to `node2nix`, I was also (or still am) a maintainer of `npm2nix`. 

You may probably not know this, but the original goal was that `node2nix` was intended to become `npm2nix` 6.0.0 (~e.g. roughly 5 years ago). 

`npm2nix` biggest problem at that time was that it couldn't cope with cyclic dependencies. It contained a few ad-hoc fixes that made some deployments with cyclic dependencies work, but in general it could not cope with it.

I realized that we needed a radically different generation approach to support cyclic dependencies (and also some of the weird version resolution quirks).  I developed a new approach from scratch and this became `node2nix`, although I was still calling my generator `npm2nix` as well.

When it became feature complete, most people didn't object, and 1 community member was heavily in favor of making it the new npm2nix 6.0.0, I also got some fierce resistance by a few, because of some of the opinionated changes that I made.

As a result, I have decided to rename my project `node2nix` (the name actually sounds pretty bad), and both tools co-existed for a quite while. So technically, while `node2nix` was originally developed as a replacement, it became an alternative.

At some point, the contributors to `npm2nix` stopped working on it, and we ran into cyclic dependency problems again in Nixpkgs.

At some point, I decided to propose `node2nix` as a replacement, and that's how it became the generator in Nixpkgs.

**Comment by Ericson2314** [comment]
@svanderburg Thanks for the history there! So it sounds like this point this is definitely the uncontested successor? (At least unless unless one counts `yarn2nix` as competitor.)

**Comment by chaoflow** [comment]
@svanderburg May I ask about the status of this project, are you still maintaining it?

**Comment by svanderburg** [comment]
@chaoflow yes, but the biggest problem is that many of the current issues are the symptoms of fundamental problems for which no proper quick fixes exist. node2nix is basically running into quite a few limitations to support nowadays' use cases. There are pull requests that fix some issues in some way, and others may actually make thing worse. IMO, most of these issues can only be properly solved by doing some substantial revisions to the whole architecture.

I've applied merges in the past, but working around the design limitations is making things more and more difficult.

The good news is that I'm already working on solving these major problems for a while, but unfortunately, they require some fundamental re-engineering work. To name a few:

* A lot of problems stem from the fact that the build environment: `node-env.nix` lacks certain important features, it suffers from problems with recursion and it has evolved into a very complex implementation that is barely maintainable, making it really hard to overcome the current limitations.
* The amount of NPM packages in Nixpkgs is quite big, and results in all kinds of scalability issues -- it takes a long time to regenerate packages, and it doesn't fit well within the package contribution workflow. I have ideas how we can solve this, but there are some challenges that we need to overcome
* The dependency resolution algorithm no longer works properly with newer versions of NPM. Fortunately, for most projects the differences are subtle, but it can only be solved by a rewrite of the dependency resolution algorithm. Most notable, it does not take the origins of dependencies into account -- in newer versions of NPM there is difference between a dependency downloaded from the NPM registry and an arbitrary HTTP location (although there output hashes may be the same).
* Certain use cases are not supported, and also because of its problematic design difficult to implement: for example, the ability to deploy remote development projects and running node2nix inside mono repos/hybrid development projects.
* There are certain changes in NPM with regards to handling directory dependencies -- now they must be symlinks. Again, this is very difficult to modify in `node-env.nix` and tedious to get right in the current implementation.

Although this isn't very visible, to solve the above issues I've already been working on a new architecture for package generation in the last two months. 

The good news is that I'm quite close to completing an important revision major step -- I have a replacement for the problematic `node-env.nix` module that looks very promising. The idea is that instead of a Nix expression that produces generated shell code, the installation/processing bits are delegated to a companion tool that takes care of populating the dependency tree. This new approach has the following advantages:
* It is more decently written and much more easily adjustable. The companion tool has a nice organization to conveniently traverse package lock files and perform modifications
* The resulting nodeEnv functions are much smaller and more understandable. Furthermore, they properly follow Nix conventions for phases and overrides etc. So adjusting builds with fixes etc. should be a lot easier
* The tool can also be used as a build input in other projects to make `npm install`. For example, it can be used as utility in monorepos/hybrid projects. This should address an important functional shortcoming.

I hope to report about this new companion tool in the next week or so.

So basically to summarize: I'm aware of the issues, but unfortunately a large category of them can only be solved by some major revisions to node2nix's architecture, but I'm working on it.

About NPM packages in Nixpkgs: I also think that is a discussion on its own. On Discourse, I've also proposed a different kind of approach that may be actually be more useful for Nixpkgs: https://discourse.nixos.org/t/future-of-npm-packages-in-nixpkgs/14285/10

With node2nix my aim was always to aim for accuracy. In the beginning my ideas were considered controversial (now it seems that people take them for granted), and as a result I named it node2nix. npm2nix and node2nix co-existed for a while.

For Nixpkgs, for example, if we can implement an incremental updater, we can actually live with a different and maybe slower generation process.

I hope this clarifies a bit about the direction I want to go

**Comment by svanderburg** [comment]
So basically to summarize: it's in a design overall. It will take a bit of time to complete all my planned revisions as they are not very trivial, but an important first step is coming very soon.

**Comment by chaoflow** [comment]
@svanderburg Thank you very much for detailed updated and I'm glad you're working on this!
